internal/apis/certmanager/validation/issuer.go

```diff
@@ -262,7 +262,80 @@ func ValidateVaultIssuerConfig(iss *certmanager.VaultIssuer, fldPath *field.Path
 		el = append(el, field.Invalid(fldPath.Child("caBundleSecretRef"), iss.CABundleSecretRef.Name, "specified caBundleSecretRef and caBundle cannot be used together"))
 	}
 
-	// TODO: add validation for Vault authentication types
+	el = append(el, ValidateVaultIssuerAuth(&iss.Auth, fldPath.Child("auth"))...)
+
+	return el
+}
+
+func ValidateVaultIssuerAuth(auth *certmanager.VaultAuth, fldPath *field.Path) field.ErrorList {
+	el := field.ErrorList{}
+
+	unionCount := 0
+	if auth.TokenSecretRef != nil {
+		unionCount++
+		if len(auth.TokenSecretRef.Name) == 0 {
+			el = append(el, field.Required(fldPath.Child("tokenSecretRef", "name"), ""))
+		}
+		if len(auth.TokenSecretRef.Key) == 0 {
+			el = append(el, field.Required(fldPath.Child("tokenSecretRef", "key"), ""))
+		}
+	}
+
+	if auth.AppRole != nil {
+		unionCount++
+		if len(auth.AppRole.Path) == 0 {
+			el = append(el, field.Required(fldPath.Child("appRole", "path"), ""))
+		}
+		if len(auth.AppRole.RoleId) == 0 {
+			el = append(el, field.Required(fldPath.Child("appRole", "roleId"), ""))
+		}
+		if len(auth.AppRole.SecretRef.Key) == 0 {
+			el = append(el, field.Required(fldPath.Child("appRole", "secretRef", "key"), ""))
+		}
+	}
+
+	if auth.Kubernetes != nil {
+		unionCount++
+
+		if len(auth.Kubernetes.Path) == 0 {
+			el = append(el, field.Required(fldPath.Child("kubernetes", "path"), ""))
+		}
+		if len(auth.Kubernetes.Role) == 0 {
+			el = append(el, field.Required(fldPath.Child("kubernetes", "role"), ""))
+		}
+
+		kubeCount := 0
+		if auth.Kubernetes.SecretRef != nil {
+			kubeCount++
+			if len(auth.Kubernetes.SecretRef.Name) == 0 {
+				el = append(el, field.Required(fldPath.Child("kubernetes", "secretRef", "name"), ""))
+			}
+			if len(auth.Kubernetes.SecretRef.Key) == 0 {
+				el = append(el, field.Required(fldPath.Child("kubernetes", "secretRef", "key"), ""))
+			}
+		}
+
+		if auth.Kubernetes.ServiceAccountRef != nil {
+			kubeCount++
+			if len(auth.Kubernetes.ServiceAccountRef.Name) == 0 {
+				el = append(el, field.Required(fldPath.Child("kubernetes", "serviceAccountRef", "name"), ""))
+			}
+		}
+
+		if kubeCount == 0 {
+			el = append(el, field.Required(fldPath.Child("kubernetes"), "please supply one of: secretRef, serviceAccountRef"))
+		}
+		if kubeCount > 1 {
+			el = append(el, field.Forbidden(fldPath.Child("kubernetes"), "please supply one of: secretRef, serviceAccountRef"))
+		}
+	}
+
+	if unionCount == 0 {
+		el = append(el, field.Required(fldPath, "please supply one of: appRole, kubernetes, tokenSecretRef"))
+	}
+	if unionCount > 1 {
+		el = append(el, field.Forbidden(fldPath, "please supply one of: appRole, kubernetes, tokenSecretRef"))
+	}
```

Comment:
The "at most one of" was not previously checked in [the controller code](https://github.com/cert-manager/cert-manager/blob/8444b9c101e92730fb5b454c688c03d84a8fecec/internal/vault/vault.go#L173-L207).

Adding the validation now will break anyone who previously had Issuer resources misconfigured. Users that have a misconfigured Issuer will see it in two different places:

- ~The cert-manager logs will start showing errors trying to update the Issuer spec~ Oops, my bad, we never update the Issuer spec or metadata. We only update the Issuer itself.
- The user won't be able to apply the manifest (or, say, ArgoCD will start failing).

I'll bring that up at this morning's standup.

Reply:
I updated the validation, PTAL @maelvls